### DISCUSSION TITLE
need help building modified shim

### DISCUSSION THREAD

**Discussion by vernal-ria0** [3kh0/ext-remover#482] [Help JOIN TITANIUM NETWORK TO SHARE PLEASE] [answered]

**Comment by 3kh0** [comment, marked as answer]
Are you using some kind of proxy? use a home computer, also see if you can send console logs, we cant hack your computer to see what is wrong

**Reply by vernal-ria0** [reply]
I have used a home computer. not using a proxy but I am on windows. it just does not do anything when I click it. sorry if I sound retarded

**Reply by vernal-ria0** [reply]
should I use a vm or is it not the problem

**Reply by 3kh0** [reply]
just do it once, no need to do it over and over, and wait, even if nothing shows, give it about 5 minutes with both of the screens being black before calling it quits

**Comment by vernal-ria0** [comment]
[Screen recording 2023-02-02 6.39.27 PM.webm](https://user-images.githubusercontent.com/116777371/216499535-0687f376-c2f4-4469-8844-6954ac0508b2.webm)

**Reply by vernal-ria0** [reply]
thas what happenes

**Reply by 3kh0** [reply]
just do it once, no need to do it over and over, and wait, even if nothing shows, give it about 5 minutes with both of the screens being black before calling it quits

**Comment by vernal-ria0** [comment]
maybe im doing something wrong

**Reply by 3kh0** [reply]
just do it once, no need to do it over and over, and wait, even if nothing shows, give it about 5 minutes with both of the screens being black before calling it quits

**Comment by amelia-bartlett** [comment]
## A guide to locally patching and flashing a RMA shim from another PC

If you have access to another PC, id boot a linux live USB and build it on that (clone the [sh1mmer repo](https://github.com/CoolElectronics/sh1mmer) and follow the instructions in [its README.md](https://github.com/CoolElectronics/sh1mmer/blob/beautifulworld/README.md))

### WARNING: MOST (purpose-built) LINUX LIVE USB's ARE CONFIGURED TO STORE ALL DATA IN A RAMDISK (i.e. any changes not saved to another drive will be lost) UNLESS PERSISTENCE PARTITIONS ARE CONFIGURED, THIS ALSO MEANS THERE IS A REAL RISK OF RUNNING OUT OF RAM DUE TO THE SIZE OF THE SHIM IMAGES (ESPECIALLY THE PATCHED ONES) AND YOUR BACKUP IMAGE!

Note that these instructions work on MacOS (the shim build process differs, see the sh1mmer instructions for relevant info)

Note that you will need a second USB (or (micro)SD Card, you can substitute USB for (micro)SD Card in this guide to the same effect) 

Once you have your patched RMA shim image, run `lsblk` to list all disks connected.
![image](https://user-images.githubusercontent.com/44680463/217896347-b7de1a8a-7ba7-45cc-a90b-a417ae027694.png)
Plug in the USB you want to write the shim to.  Wait a couple of seconds, then run `lsblk` again.
![image](https://user-images.githubusercontent.com/44680463/217896582-7ccdd884-6570-475a-859f-39f025cf1751.png)
This will tell you what device file is assigned to the USB you want to write the RMA shim to (it will only show up from the second `lsblk`).  You will see multiple new entries - this is fine, as some of them are partitions.  Out of the new entries, pick the one that seems to be at the root of a tree.
![image](https://user-images.githubusercontent.com/44680463/217897189-cdf36778-b1e3-4141-8a2f-5fc39a834cec.png)
(If you are having trouble finding which one it is in the new ones, it will have the shorter filename)

You now know the name of the device (in my case `sdb`).  The file path will be `/dev/[name]` (in my case it would be `/dev/sdb`)

Before proceeding further, you should probably check what filesystem is already on your USB if you want to revert it.  You should also backup any files on there (you can do this by reverting the `if=` and `of=` arguments of `dd` (pursuant to the warning given below) and replacing the `of=` argument with `of=/path/to/your/backup/image.bin` (make sure this file is not on your USB and that it does not already exist to avoid data loss).  You will get a command like this:
```shell
dd if=/dev/[name] of=/path/to/your/backup/image.bin status=progress
```
that you can run, pursuant to the below warnings and info (up until the bit about running `sync`, this is 

Now you can proceed with the writing of the USB.

### CAUTION: INCORRECT USAGE OF `dd` WILL ALMOST CERTAINTLY DAMAGE YOUR FILES, AND IN SOME CASES OTHER SYSTEM COMPONENTS!  ENSURE YOUR `if=` AND `of=` ARGUMENTS ARE CORRECT!  ANY DATA IN THE TARGET FILE (this will mean the USB stick you are writing to in this case, since on Linux, everything is considered a file) WILL BE OVERWRITTEN AND CANNOT BE RECOVERED!

Write it to the USB with:
(note you _can_ use relative paths afaik, but I would suggest using absolute paths to be sure)
(note that /dev/[name] is the file path you got previously)
(note it will likely fail with a permissions error.  prefix the command with `sudo ` and retry to counter this)
```shell
dd if=/full/path/to/your/patched/rma/shim.bin of=/dev/[name] status=progress
```
(If you forget the `status=progress` then its fine, you just won't get any feedback from the command.  It will still be running, be patient and DO NOT UNPLUG ANYTHING!)

Wait for the command to exit and drop you back at a shell.  Run `sync` to make sure the disk cache is flushed (wait for this to finish) before unplugging your USB.

You now have a patched RMA shim image on a USB ready to be used, congrats!

Once you are done, you can run:
```shell
dd if=/path/to/your/backup/image.bin of=/dev/[name] status=progress
```
to get your data back to your USB.  Remember to run `sync` when done!

### Wait!  How do I get my USB stick back?  I didn't create a backup image!

Don't worry, your USB stick is not read-only and can be reverted to act as a 'normal' USB stick rather than a sh1mmer shim.  However, step-by-step instructions on doing so are not in the scope of this guide.

### CAUTION: ENSURE YOU SELECTED THE RIGHT DISK!  IMPROPER USAGE OF A PARTITION EDITOR WILL CAUSE DATA LOSS!

You should delete all the partitions on your USB disk in a partition editor (e.g. KDE Partition Manager, GParted, macOS X Disk Utility, whatever the Windows one was) or just delete the partition table.  Then, create a new partition spanning the entire disk of the appropriate filesystem (if unsure, fat32 is usually used on external disks).  **REMEMBER TO HIT 'APPLY' (or equivalent) BEFORE EXITING!**

**Reply by amelia-bartlett** [reply]
Someone also took a backup of the prebuilt shims, although I cannot vouch for their authenticity.

https://github.com/DiffuseHyperion/sh1mmer-archived#new-description
